org.eclipse.scanning.api/src/org/eclipse/scanning/api/database/IExperimentDatabaseService.java

```diff
@@ -0,0 +1,190 @@
+     * @throws IllegalArgumentException
+     * @throws Exception
+     */
+    <T> Future<Id> insert(T action, boolean blocking)  throws IllegalArgumentException, Exception;
+ 
+    /**
+	 * The method can be used to create new records for instance:
+	 
+	 <code><pre>
+	    DataCollectionGroup grp = new DataCollectionGroup();
+		grp.setProposalCode("cm");
+		grp.setProposalNumber(14451);
+		grp.setSessionNumber(1);
+		grp.setSampleId(11550L);
+		Long dataCollectionGrpId = service.upsert(grp, true);
+	 </pre></code>
+		
+	   This method can be used to update records for instance:
+	 <code><pre>
+	   	DataCollectionGroup grp = new DataCollectionGroup();
+		grp.setId(dataCollectionGrpId); // We know the id which we are upserting
+		grp.setComments("Hello World");
+		long code = service.upsert(grp, false); // Add the job to a queue to be processed later.
+	 </pre></code>
+
+	 * 
+	 * @param entry
+	 * @param blocking - true if operation should be blocking. If false 
+	 * @return upsert code or -1 if the upsert was asynchronous (non-blocking)
+     * @throws IllegalArgumentException
+     * @throws Exception
+	 */
+    <T> Future<Id> upsert(T entry, boolean blocking) throws IllegalArgumentException, Exception;
+	
+	/**
+	 * The method cannot be used to create new records.		
+	   This method can be used to update records for instance:
+	 <code><pre>
+		DataCollectionExperiment experiment = new DataCollectionExperiment();
+		experiment.setId(eid); // We know the id which we are updating
+		experiment.setSlipGapVertical(0.01);
+		long code = service.upsert(grp, false); // Add the job to a queue to be processed later.
+	 </pre></code>
+
+	 * 
+	 * @param entry
+	 * @param blocking - true if operation should be blocking. If false 
+	 * @return upsert code or -1 if the upsert was asynchronous (non-blocking)
+     * @throws IllegalArgumentException
+     * @throws Exception
+	 */
+    <T> Future<Id> update(T entry, boolean blocking) throws IllegalArgumentException, Exception;
+	
+   
+    /**
+     * This action allows several operations to be completed in one action.
+     * The other operations allow one or more actions of the same type but 
+     * this 'composite' operation allows one or more upsert/update/insert
+     * to be completed. The order of processing is insert/upsert/update 
+     * for a composite operation list because this allows an insert and
+     * update of a record (similar to upsert!).
+     * 
+     * T in the case of this operation will normally be a CompositeBean.
+     * There are currently no other implementations of the Composite but
+     * we have decided not to hard code the required type in order to match
+     * the other methods.
+     * 
+     * @param action
+     * @param blocking
+     * @return
+     * @throws IllegalArgumentException
+     * @throws Exception
+     */
+    <T> Future<Id> composite(T action, boolean blocking)  throws IllegalArgumentException, Exception;

```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS00112)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the declaration of thrown exception 'java.lang.IllegalArgumentException' which is a subclass of 'java.lang.Exception'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)